### PR TITLE
cmake: WITH_SPDK=ON by default

### PR DESCRIPTION
**Pull request by tchaikov**

### CODE REVIEW

File: debian/control [hunk 1]
```diff
@@ -29,6 +29,7 @@ Build-Depends: bc,
                libbabeltrace-ctf-dev,
                libbabeltrace-dev,
                libblkid-dev (>= 2.17),
+               libcunit1-dev,
```
Comment (tchaikov) [inline debian/control]: @liupan1111 this what i mean by `debian/control`.

File: ceph.spec.in [hunk 2]
```diff
@@ -183,6 +183,7 @@ BuildRequires:	keyutils-libs-devel
 BuildRequires:	libibverbs-devel
 BuildRequires:  openldap-devel
 BuildRequires:  openssl-devel
+BuildRequires:  CUnit-devel
```
Comment (tchaikov) [inline ceph.spec.in]: @smithfarm does this work for SLE? or do you want to WITH_SPDK=ON for the build also?

File: cmake/modules/BuildDPDK.cmake [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    set(machine "native")
+  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "arm|ARM")
+    set(arch "arm")
+    set(machine "armv7a")
+  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|AARCH64")
+    set(arch "arm64")
+    set(machine "armv8a")
+  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "(powerpc|ppc)64")
+    set(arch "ppc_64")
+    set(machine "power8")
```
Comment (tchaikov) [inline cmake/modules/BuildDPDK.cmake]: @smithfarm do you think we should enable DPDK on "power8" also? as http://dpdk.org/doc/guides/linux_gsg/build_dpdk.html claims that DPDK supports power8 as well.

File: cmake/modules/BuildDPDK.cmake [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}/src/spdk/dpdk
+    OUTPUT_VARIABLE supported_targets)
+
+  if(NOT supported_target MATCHES ${target})
+    message(FATAL "not able to build DPDK support: "
+      "unsupported target \"${target}\"")
+  endif()
+
+  if(WITH_KERNEL_DIR)
+    if(NOT IS_DIRECTORY ${WITH_KERNEL_DIR})
```
Comment (tchaikov) [inline cmake/modules/BuildDPDK.cmake]: @liupan1111 this is my biggest concern. the target machine's kernel is not always the same as the build host. even we managed to fulfill the build-requirement by installing the package of `linux-headers-\`uname -r\``. there is good chance that the built DPDK does not work for the deployed machine.

@liewegas what do you think?

Comment (liewegas) [reply cmake/modules/BuildDPDK.cmake]: I'm not sure I understand the kernel version dependency.  Does the build have to be built against a specific kernel?  I.e., if we build on the centos 7.3 kernel will the built resuilt break on 7.4?  Or a build on the 16.04 kernel break on a later HWE kernel? 

As long as those scenerios behave okay, I think it's fine to continue.  If there is a hard kernel dependency, then I'm not sure what we should do.. that would suck.  :(

**Comment by smithfarm** [comment]
@tchaikov As usual, I'm running behind. . . I haven't tried to build master on SLE lately. I will try building this branch and let you know the result.

File: cmake/modules/BuildDPDK.cmake (hunk 4 above)
Comment (smithfarm) [reply cmake/modules/BuildDPDK.cmake]: For SLE, I'm pretty sure kernel module code needs to be integrated into, built as part of, and shipped in the kernel package itself. That may be an oversimplification, but certainly I wouldn't expect to be building kernel module packages via `ceph.spec`.

Comment (smithfarm) [reply cmake/modules/BuildDPDK.cmake]: @ddiss Can you vet my comment, above?

Comment (tchaikov) [reply cmake/modules/BuildDPDK.cmake]: yeah, to my best knowledge, kernel modules are not portable across different versions. anyway, i am disabling the kernel modules in DPDK. 

@liupan1111 could you help test the build w/o the kernel modules? thanks

File: cmake/modules/patch-dpdk-conf.sh [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+setconf CONFIG_RTE_LIBRTE_MBUF_OFFLOAD n
+# Disable unmaintained features
+setconf CONFIG_RTE_LIBRTE_POWER n
+
+setconf CONFIG_RTE_EAL_IGB_UIO n
+setconf CONFIG_RTE_LIBRTE_KNI n
+setconf CONFIG_RTE_KNI_KMOD n
+setconf CONFIG_RTE_KNI_PREEMPT_DEFAULT n
+
+setconf CONFIG_RTE_APP_EVENTDEV n
```
Comment (tchaikov) [inline cmake/modules/patch-dpdk-conf.sh]: i know, this option is not in the used DPDK yet.

File: cmake/modules/patch-dpdk-conf.sh [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        echo $key=$val >> ${conf}
+    fi
+}
+
+conf=$1
+shift
+machine=$1
+shift
+
+setconf CONFIG_RTE_MACHINE "${machine}"
```
Comment (tchaikov) [inline cmake/modules/patch-dpdk-conf.sh]: @yuyuyu101 do you think we can have a functional SPDK backend without all these kernel modules?

**Comment by tchaikov** [comment]
@smithfarm shaman is happy. could you give the branch a spin?

File: CMakeLists.txt [hunk 7]
```diff
@@ -348,6 +348,10 @@ endif()
 
 option(WITH_DPDK "Enable DPDK messaging" OFF)
 if(WITH_DPDK)
+  if(NOT USE_CRYPTOPP)
+    message(FATAL_ERROR "CRYPTOPP must be supported when enable DPDK.")
```
Comment (tchaikov) [inline CMakeLists.txt]: the the incompatible option settings at the toplevel cmake script.

File: src/CMakeLists.txt [hunk 8]
```diff
@@ -687,6 +657,32 @@ if(HAVE_ARMV8_CRC)
   list(APPEND ceph_common_deps common_crc_aarch64)
 endif(HAVE_ARMV8_CRC)
 
+if(WITH_DPDK)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: so we can have the `WITH_DPDK` support in a single place.

File: src/CMakeLists.txt [hunk 9]
```diff
@@ -687,6 +657,32 @@ if(HAVE_ARMV8_CRC)
...
+  add_library(common_async_dpdk STATIC
+    ${common_async_dpdk_files})
+  # to ready the dpdk header files
+  add_dependencies(common_async_dpdk
+    ${DPDK_LIBRARIES})
+  list(APPEND ceph_common_deps common_async_dpdk)
+  target_link_libraries(common_async_dpdk
+    ${DPDK_LIBRARIES})
+  set_target_properties(common_async_dpdk PROPERTIES
+    COMPILE_FLAGS "-march=native -I${DPDK_INCLUDE_DIR}")
```
Comment (tchaikov) [inline src/CMakeLists.txt]: so we can avoid adding "-march=native" globally. this will be a show stopper, if we want to enable DPDK by default.

**Comment by tchaikov** [comment]
retest this please.

**Comment by liu-chunmei** [comment]
@tchaikov does your PR support the following 3 cases: 
1. only enable WITH_DPDK = ON
2. only enable WITH_SPDK = ON
3. enable both WITH_DPDK =ON and WITH_SPDK =ON
Is there build error for these 3 scenarios? and for these 3 scenarios,  can the ceph system start up? 
if yes, I think we can merge the patch first. Then for the functional issue we can fix later.

**Comment by liu-chunmei** [comment]
@tchaikov Maybe you need squash the several commits into one before merge it.

**Comment by tchaikov** [comment]
@liu-chunmei i squashed the two DPDK cleanup changes into a single one. but i'd like to keep other changes separated commits. as in this way, it'd help the posterity to understand changes.

**Pull request merged**